toolchain/check/function.cpp

```diff
@@ -0,0 +1,216 @@
+  if (fn.return_type_id == SemIR::TypeId::Error) {
+    return true;
+  }
+  for (auto param_refs_id : {fn.implicit_param_refs_id, fn.param_refs_id}) {
+    if (param_refs_id != SemIR::InstBlockId::Empty) {
+      for (auto param_id : context.inst_blocks().Get(param_refs_id)) {
+        if (context.insts().Get(param_id).type_id() == SemIR::TypeId::Error) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Returns true if a param agrees. The caller is expected to provide a
+// diagnostic.
+// TODO: Consider moving diagnostics here, particularly to differentiate
+// between type and name mistakes. For now, taking the simpler approach because
+// I also think we may want to refactor params.
+static auto DoesParamAgree(Context& context, SemIR::InstId new_param_ref_id,
+                           SemIR::InstId prev_param_ref_id) -> bool {
+  auto new_param_ref = context.insts().Get(new_param_ref_id);
+  auto prev_param_ref = context.insts().Get(prev_param_ref_id);
+  if (new_param_ref.kind() != prev_param_ref.kind() ||
+      new_param_ref.type_id() != prev_param_ref.type_id()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AddrPattern>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AddrPattern>().inner_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AddrPattern>().inner_id);
+  }
+
+  if (new_param_ref.kind() != prev_param_ref.kind()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AnyBindName>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AnyBindName>().value_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AnyBindName>().value_id);
+  }
+
+  CARBON_CHECK(new_param_ref.Is<SemIR::Param>()) << new_param_ref;
+  CARBON_CHECK(prev_param_ref.Is<SemIR::Param>()) << prev_param_ref;
+
+  auto new_param = new_param_ref.As<SemIR::Param>();
+  auto prev_param = prev_param_ref.As<SemIR::Param>();
+  return new_param.name_id == prev_param.name_id;
+}
+
+// Returns true if two param refs agree.
+static auto DoParamAgrees(Context& context, SemIR::InstId new_decl_id,
+                          SemIR::InstBlockId new_param_refs_id,
+                          SemIR::InstId prev_decl_id,
+                          SemIR::InstBlockId prev_param_refs_id,
+                          llvm::StringLiteral param_diag_label) -> bool {
+  // This will often occur for empty params.
+  if (new_param_refs_id == prev_param_refs_id) {
+    return true;
+  }
+  const auto new_param_ref_ids = context.inst_blocks().Get(new_param_refs_id);
+  const auto prev_param_ref_ids = context.inst_blocks().Get(prev_param_refs_id);
+  if (new_param_ref_ids.size() != prev_param_ref_ids.size()) {
+    CARBON_DIAGNOSTIC(
+        FunctionSignatureParamCountDisagree, Error,
+        "Function declaration with {0}{1} parameter(s) disagrees.", int32_t,
```

Comment:
I think our choice of terminology isn't helping us present a clear diagnostic here. Maybe we can massage it a bit:

"Function redeclaration with {0}{1} parameter(s) does not agree with previous declaration."
"Previous matching declaration has {0}{1} parameter(s)."

... but really the match / agree terminology doesn't seem to be working well here. :(

If we keep the current terminology, I think we should always be saying "[dis]agrees *with something*" not just abstractly "[dis]agrees" -- it should be a relation between two things.

[I also really don't like the `(s)`, but that's not new in this PR.]

Reply:
Rephrasing, avoiding `(s)`